components/Init/src/virtio_con.c

```diff
@@ -0,0 +1,175 @@
+/*
+ * Copyright 2022, UNSW (ABN 57 195 873 179)
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <stdint.h>
+#include <string.h>
+#include <autoconf.h>
+
+#include <utils/util.h>
+
+#include <camkes.h>
+#include <camkes/dataport.h>
+
+#include <sel4vmmplatsupport/drivers/virtio_con.h>
+#include <sel4vmmplatsupport/device.h>
+
+#include <sel4vm/guest_irq_controller.h>
+#include <sel4vm/boot.h>
+
+#include <platsupport/serial.h>
+
+#include <virtqueue.h>
+#include <camkes/virtqueue.h>
+
+#include "virtio_con.h"
+#include "virtio_irq.h"
+
+/*
+ * This file is part of the virtio console backend, it's responsible for
+ * sending/receiving data to/from other VM's through using our virtqueue
+ * implementation.
+ */
+
+static virtio_con_t *virtio_con;
+
+typedef struct virtio_con_cookie {
+    virtio_con_t *virtio_con;
+    vm_t *vm;
+} virtio_con_cookie_t;
+
+/* Configured in camkes */
+#define NUM_PORTS ARRAY_SIZE(serial_layout)
+
+/* 4088 because the serial_shmem_t struct has to be 0x1000 bytes big */
+#define BUFSIZE (0x1000 - 2 * sizeof(uint32_t))
+
+/* This struct occupies exactly 1 page and represents the data in the shmem region */
+typedef struct serial_shmem {
+    uint32_t head;
+    uint32_t tail;
+    char buf[BUFSIZE];
+} serial_shmem_t;
```

Comment:
Please expand the comment to say what head and tail do, and that access can be lock free (in what circumstances?) and that head and tail increase continually, are used mod BUFSIZE to index into buf

Reply:
I'd prefer a different way of indexing.
  Always add at buf[++head % BUFSIZE]
  (typically add by memcpying into buf, then a memory barrier, then update head, so consumer can't see until all is added)
  Always take from [tail++ % BUFSIZE]
  empty = (head-tail)%BUFSIZE==0
  full = (head - tail) % BUFSIZE == BUFSIZE-1

Reply:
Right now you have the invariants, 0 <= head < BUFSIZE
0 <= tail < BUFSIZE
empty = head - tail == 0
full = head > tail ? head - tail == BUFSIZE-1 : head + BUFSIZE-tail == BUFSIZE-1

Reply:
What is the advantage of having ringbuffer indexes that increase continually? Seems like it can make the calculation easier, but you need to do a modular every time when accessing the buffer.

Reply:
Also, the current structure of the ringbuffer is originally used by the [SerialServer](https://github.com/seL4/global-components/blob/master/components/SerialServer/src/serial.c#L42).
The arm version of the virtio console backend (see [here](https://github.com/seL4/camkes-vm/blob/ce6065f3caa230ef6c712e61190daebeb74fc005/components/VM_Arm/src/modules/virtio_con.c#L146)) needs to handle VM-VM transactions as well as VM-Serial transactions, and has only one single interface for both of them. 
I can only change the VM-VM ringbuffers to use the indexing you propose, but not the VM-Serial one.  Not sure if it would be a good idea to have two variants of ringbuffer in the same component as they need to be handled differently.

Reply:
> What is the advantage of having ringbuffer indexes that increase continually? Seems like it can make the calculation easier, but you need to do a modular every time when accessing the buffer.

Main thing is it's much easier to check the invariants, and it's a lot easier to get right.  The corner cases otherwise keep having to check for overflow; generally speaking a branch structure is harder to check for correctness, and slower than a simple modulo.  That's even more true if we could make the BUFSIZ a power of two,when modulo becomes an 'and'